// @tqrsa/kyber/index.js
/*/////////////////////////////////////////////////////////////////////////////

  Author  : @qodeninja

/////////////////////////////////////////////////////////////////////////////*/

    const mr = Math.random, 
          mf = Math.floor, 
          tl='toLowerCase',
          tu='toUpperCase',
          isa=Array.isArray,
          sfc=String.fromCharCode,
          log=console.log;

/*/////////////////////////////////////////////////////////////////////////////
// GX Lib
/////////////////////////////////////////////////////////////////////////////*/


    const timestamp    = ()   => { return +new Date(); }
    const stringify    = (s)  => { return JSON.stringify(s) + ""; }
    const keyParse     = (u)  => { let q=Object.keys(u); return(q.length ? q : '--'); }
    const quickRand    = (m,n)=> { n=n||0;m=(isa(m)?m.length:m); return mf(mr()*m)+n; }
    const hexGen       = (n)  => { function b(a){ return a?(a^mr()*16>>a/4).toString(16):([1e7]+1e3+4e3+'').replace(/[018]/g,b); } return b(n); }
    const clockGen     = (m,c)=> { c=c||1;return (x)=>{ if(m!==null && c<=m){ if(x) c=0;return c++;} return null;}};
    const counter      = clockGen(true);

    const alnumGen = (l,a)=>{
      a=a&&a[tl](),s="",i=0,m=a=="a"?10:0,n=a=="n"?10:62;
      for(;i++<l;){let r=mr()*(n-m)+m<<0; s+=sfc(r+=r>9?r<36?55:61:48);}
      return s;
    }

    const kvGen = (f,i=0) =>{     
      let len = f||quickRand(5,1);
      let s = alnumGen(len,'a')[tl]();
      s = s.split('');
      clk = clockGen(s.length);
      dd = usDateRangeGen( toDate([3,22,1905]), toDate([3,22,2015]) );
      let o = { row : i, id : alnumGen(len,'n'), name : nameGen(), bday : dd()  };
      while(s.length >= 1) o[`key_${clk()}`] = `val_${s.pop()}`;
      return o;
    }

    const dataGen = (x,f) => {
      var res=[],i=0,c=clockGen(x),f = f||quickRand(5,1);
      while(i!==null){ i = c(); if(i==null) continue; res.push(kvGen(f,i)); } 
      return res;
    }

    const nameGen = (dict) => {
      i=0, j=0, fl='', nl='', dict = dict || { 
        pre:[ 'tom', 'ark', 'far', 'shi', 'zag', 'mar', 'dav', 'jen', 'rob', 'bren', 'sam' ],
        suf:[ 'is', 'malo', 'zak', 'ben', 'wonk', 'quez', 'ender', 'ali', 'son', 'kor' ]
      }
      fnl = clockGen(1);
      lnl = clockGen(quickRand(3));
      while(i!==null){ i = fnl(); fl +=dict.pre[quickRand(dict.pre)] + dict.suf[quickRand(dict.suf)]; } 
      while(j!==null){ j = lnl(); nl +=dict.suf[quickRand(dict.suf)]; } 
      function tcase(a){ return a.charAt(0)[tu]() + a.substr(1)[tl]();}
      var name = tcase(fl)+" "+tcase(nl);
      return (name);
    }



    const dateRangeGen = (s,e,cb) => {
      return ()=>{ d=new Date(s.getTime() + Math.random() * (e.getTime() - s.getTime())); if(cb) return cb(d); return d; }
    }

    const toDate = (d) => {
      if(isa(d)) return new Date(d[2],d[1],d[0]);
      return new Date();
    }

    const usDate = (d) => { d = new Date(d); return d.toLocaleDateString('en-US'); }

    const usDateRangeGen = (s,e, cb) => {
      return dateRangeGen(s,e,usDate);
    }


    //usDateGenerator =>  var dd = fx.dg( fx.td([3,22,1900]), fx.td(), locale );
    //dateGen(new Date(2012, 0, 1), new Date())

    //SuperGenerator => generator  => data
    const superGenerator = ( api, conf, input, dx ) =>{

    }

    const sseGenerator = ( ms, mod, srl, generator) => {


      function intervalGenerator(o){
        to = setInterval(function(){
              console.log('writing ' + testdata);
              o.write('data: ' + JSON.stringify({ msg : testdata }) + '\n\n');
             }, ms);
        return to;
      }


      mod.get( srl, ( i,o,n )=> {

          i.socket.setKeepAlive(true);
          i.socket.setTimeout(0);

          o.writeHead(200, {
            'Connection'   : 'keep-alive',
            'Content-Type' : 'text/event-stream',
            'Cache-Control': 'no-cache'
          });

          to = intervalGenerator(o);
          mod.get('/api/timeout', ( i,o,n )=> {
            clearInterval(to);
            o.send('Killing Timer');
          });

          o.on('close', ()=> {
            clearInterval(to);
          });


      });


    }

/*/////////////////////////////////////////////////////////////////////////////
// GX Driver
/////////////////////////////////////////////////////////////////////////////*/

  module.exports = {
    t : timestamp,
    j : stringify,
    k : keyParse,
    c : counter,
    r : quickRand,
    h : hexGen,
    a : alnumGen,
    d : dataGen,
    cg : clockGen,
    dg : dateRangeGen,
    td : toDate,
    ud : usDate,
    udg: usDateRangeGen,
    ng : nameGen,
    kv : kvGen,
    sse: sseGenerator
  };
